### PR TITLE
sql: timestamp should accept DB2 formatted timestamps

### LINKED ISSUE #49768: sql: timestamp should accept DB2 style timestamps
DB2 timestamps use the following format:
2014-03-15-04.38.53.399853

Which differs from ours:
2014-03-15 04:38:53.399853

When trying to import data directly from a db2 dump file requires some pre-processing.  The DB2 format seems fairly straightforward and I can see no reason why we can't make this a valid insert format as well.

https://www.db2tutorial.com/db2-basics/db2-timestamp/

### PR DESCRIPTION
**Pull request by sooryranga**
fixes : #49768

This PR adds additional extracting field logics that
allow db2 formatted timesstamps to be parsed.

Release note (sql change): The pgdate util can now parse
db2 formatted timestamps. This can be used to import db2 dumps.

### CODE REVIEW

**Comment by irfansharif** [comment]
@jordanlewis: feel free to re-assign, not sure who's the expert here now.

**Review by otan — COMMENTED** [review]
Thanks for contributing! I've got a few questions with regards to how it affects existing cases, as well as a request for more tests.

Can you also add a test in https://github.com/cockroachdb/cockroach/blob/d9ae34b2851843ec2c350436f9cf7add550b0273/pkg/sql/sem/tree/testdata/eval/cast#L490? `go test ./pkg/sql/sem/tree -run 'TestEval/cast'` should trigger it.

File: pkg/util/timeutil/pgdate/parsing_test.go [hunk 1]
```diff
@@ -632,6 +632,10 @@ var timestampTestData = []timeData{
 		exp:         time.Date(2000, 1, 1, 2, 2, 2, 567000000, time.FixedZone("", 9*60*60+30*60+15)),
 		hasTimezone: true,
 	},
+	{
+		s:   "2014-03-15-04.38.53.399853+07:35",
```
[the remark this answers is not in the document]
Comment (otan) [reply pkg/util/timeutil/pgdate/parsing_test.go:636]: a few error conditions, e.g. `2007-09-24-15.0` may also be nice to test.

File: pkg/util/timeutil/pgdate/field_extract.go [hunk 2]
```diff
@@ -252,6 +252,7 @@ func (fe *fieldExtract) interpretNumber(numbers []numberChunk, idx int, textMont
 		//                   ^^^^^^
 		// Example: 1999.123, a year + day-of-year.
 		//               ^^^
+		// Example: 04.04.04.913231+00:00, odb2 timestamp
```
[the remark this answers is not in the document]
Comment (Arun4rangan) [reply pkg/util/timeutil/pgdate/field_extract.go:259]: Good catch!

File: pkg/util/timeutil/pgdate/field_extract.go [hunk 3]
```diff
@@ -451,7 +459,10 @@ func (fe *fieldExtract) interpretNumber(numbers []numberChunk, idx int, textMont
 			fe.tweakYear = true
 		}
 		return fe.SetChunk(fieldYear, chunk)
-
+	case !fe.Wants(fieldDay) && fe.Wants(fieldHour) && chunk.separator == '-':
+		// Example: "YYYY-MM-DD-HH-MM-SS"
```
[the remark this answers is not in the document]
Comment (Arun4rangan) [reply pkg/util/timeutil/pgdate/field_extract.go]: No there isn't, it's a mistake.

File: pkg/util/timeutil/pgdate/parsing.go [hunk 4]
```diff
@@ -55,6 +55,8 @@ var (
 		fieldTZHour, fieldTZMinute, fieldTZSecond)
 	timeRequiredFields = newFieldSet(fieldHour, fieldMinute)
 
+	odbTimeRequiredFields = newFieldSet(fieldHour, fieldMinute, fieldSecond)
```
Comment (otan) [inline pkg/util/timeutil/pgdate/parsing.go:58]: any reason this is called `odb`?

File: pkg/util/timeutil/pgdate/field_extract.go [hunk 5]
```diff
@@ -246,12 +248,17 @@ func (fe *fieldExtract) Get(field field) (int, bool) {
...
+	if len(numbers) > idx+1 {
+		nextSep = numbers[idx+1].separator
+	}
 	switch {
 	case chunk.separator == '.':
 		// Example: 04:04:04.913231+00:00, a fractional second.
 		//                   ^^^^^^
 		// Example: 1999.123, a year + day-of-year.
 		//               ^^^
+		// Example: 04.04.04.913231+00:00, odb timestamp
```
Comment (otan) [inline pkg/util/timeutil/pgdate/field_extract.go:261]: what is `odb`?

File: pkg/util/timeutil/pgdate/field_extract.go [hunk 6]
```diff
@@ -263,6 +270,14 @@ func (fe *fieldExtract) interpretNumber(numbers []numberChunk, idx int, textMont
 			// BUT NOT: 1999 1
 			return fe.SetDayOfYear(chunk)
 
+		case fe.Wants(fieldMinute) && chunk.v <= 60 && chunk.v >= 0:
+			// db2 timestamp allows for minutes to recorded with periods.
+			// Example 13.12.50 (hh.mm.ss)
+			fe.isDB2 = true
```
Comment (otan) [inline pkg/util/timeutil/pgdate/field_extract.go:276]: should `fe.isDB2` already be set by the `-` separator below, or is the `-` between the YYYY-MM-DD and HH.MM.SS optional?

File: pkg/util/timeutil/pgdate/field_extract.go [hunk 7]
```diff
@@ -64,6 +64,8 @@ type fieldExtract struct {
 	tzSign int
 	// Tracks the fields that we want to extract.
 	wanted fieldSet
+	// Tracks whether the current timestamp formes db2 format
```
Comment (otan) [inline pkg/util/timeutil/pgdate/field_extract.go:67]: nit: formes seems out of places. "Tracks whether the current timestamp is of db2 format.", perhaps.

File: pkg/storage/enginepb/mvcc3.pb.go [hunk 8]
```diff
@@ -4207,7 +4207,9 @@ var (
 	ErrIntOverflowMvcc3   = fmt.Errorf("proto: integer overflow")
 )
 
-func init() { proto.RegisterFile("storage/enginepb/mvcc3.proto", fileDescriptor_mvcc3_9bc532bf92053320) }
```
Comment (otan) [inline pkg/storage/enginepb/mvcc3.pb.go:4210]: please revert these protobuf .pb.go changes

File: pkg/sql/sem/tree/testdata/eval/cast [hunk 9]
```diff
@@ -1070,3 +1070,48 @@ eval
...
+
+eval
+'1969-12-30-01.00.00-7'::timestamptz::text
+----
+'1969-12-30 08:00:00+00:00'
+
+eval
+time '1969-12-30-01.00.00.12345'
+----
+'01:00:00.12345'
```
Comment (otan) [inline pkg/sql/sem/tree/testdata/eval/cast:1117]: please add \n at end of file.

File: pkg/sql/sem/tree/testdata/eval/cast [hunk 10]
```diff
@@ -1070,3 +1070,48 @@ eval
 '-10'::interval::decimal
 ----
 -10.000000000
+
+eval
+'1969-12-30-01.00.00'::timestamp
```
Comment (otan) [inline pkg/sql/sem/tree/testdata/eval/cast:1075]: please put these near the other timezone cases. they do not have to be exhaustive either -- just one or two is fine (these should be unit tests instead if we need them)

**Review by otan — COMMENTED** [review]
Thanks! couple more small nits.

File: pkg/util/timeutil/pgdate/field_extract.go (hunk 6 above)
Comment (Arun4rangan) [reply pkg/util/timeutil/pgdate/field_extract.go:276]: You could have something like `yyyy-mm-dd hh.mm.ss` which is a DB2 type, we need to set DB2 on `mm` so if we get something like `yyyy-mm-dd hh.mm:ss` we could throw an error. 

In this cause it won't `-` but still should be treated as db2 timestamp.

File: pkg/util/timeutil/pgdate/field_extract.go [hunk 11]
```diff
@@ -263,6 +264,13 @@ func (fe *fieldExtract) interpretNumber(numbers []numberChunk, idx int, textMont
 			// BUT NOT: 1999 1
 			return fe.SetDayOfYear(chunk)
 
+		case fe.Wants(fieldMinute) && chunk.v <= 60 && chunk.v >= 0:
```
[the remark this answers is not in the document]
Comment (Arun4rangan) [reply pkg/util/timeutil/pgdate/field_extract.go:273]: You are right, using `isDB2` as a trigger prevents unnecessary field extraction.

**Review by otan — APPROVED** [review]
Thanks!

**Pull request merged**